### PR TITLE
Gitlab oauth pkce authorization flow for CLI (#554)

### PR DESCRIPTION
**Pull request by JamWils**
Closes: #554 

This adds a Gitlab authorization flow to the code base using the PKCE flow.  A new set of utility functions were added to generate a random string between a random length defined by min and max.  This string is generated with in a new struct type CodeVerifier.  The CodeVerifier struct contains all of the information necessary to perform a PKCE operation through oauth.

A new interface was defined called AuthFlow and an internal struct gitlabAuthFlow implements that interface.  The gitlabAuthFlow handles generating the authorization request as well as a callback handler that can be used via CLI and an assumption that it will also work via API.  A new BlockingCLIAuthHandler was also created.  This will run through the entire authorization process for Gitlab, although it can easily be adjusted for any third party service needed.  This will start a temporary server and wait for the user to use the authorization url provided.  Once the third-party service calls us back via the specified redirect uri then we fetch the access token and shutdown the temporary server.

The internal package was leveraged for things we would never want leaked out of the auth package.  The types package was created since I have found this a nice pattern to avoid cyclical imports in larger code bases.

Ginkgo had some weird behaviors when testing the CLI Auth flow.  Some unit tests are specifically separated into their own describe blocks due to weird behavior from the stubbed out round tripper.  I believe this is due to how Ginkgo leverages a single test. 

Unit tests were written to test all new functionality and edge cases.  In addition, for manual testing I created a temporary command so I could test the authorization flow through the CLI.  It worked as expected and then I removed the temporary command.

### CODE REVIEW

File: pkg/services/auth/gitlab.go [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+	"github.com/weaveworks/weave-gitops/pkg/services/auth/types"
+	"io"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+)
+
+const (
+	serverShutdownSuccessMessage = "The authorization flow was completed, you may close your browser and go back to the CLI.\n"
```
Comment (jpellizzari) [inline pkg/services/auth/gitlab.go:18]: :thinking: Assuming this is what shows up in the browser, we probably need to have a nicely-rendered page here eventually. Might be do-able in a follow up.

**Review by jpellizzari — COMMENTED** [review]
Mostly questions

File: pkg/services/auth/gitlab.go (hunk 1 above)
Comment (JamWils) [reply pkg/services/auth/gitlab.go:18]: Yeah, that would be nice.  Definitely worth a follow-up

File: pkg/utils/string_test.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+		value2, _ := GenerateRandomString(10, 20)
+		value3, _ := GenerateRandomString(10, 20)
+
+		Expect(value1).NotTo(Equal(value2))
+		Expect(value1).NotTo(Equal(value3))
+		Expect(value2).NotTo(Equal(value3))
+	})
+
+	It("panics when min is greater than max", func() {
+		Expect(func() { GenerateRandomString(20, 10) }).Should(Panic())
```
[the remark this answers is not in the document]
Comment (JamWils) [reply pkg/utils/string_test.go]: I almost returned an error, but then decided not to.  The developer should see this panic during the development phase.  If we use this function with varying lengths during runtime then we want to update it  However, I find it hard to see where we would do that.  I can change it if you want.

File: pkg/services/auth/internal/verifier.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+		value: value,
+	}, nil
+}
+
+func (c CodeVerifier) CodeChallenge() string {
+	h := sha256.New()
+	h.Write([]byte(c.value))
+
+	encodedValue := base64.StdEncoding.EncodeToString(h.Sum(nil))
+	encodedValue = strings.Replace(encodedValue, "+", "-", -1)
```
[the remark this answers is not in the document]
Comment (JamWils) [reply pkg/services/auth/internal/verifier.go:37]: Nope url will include them as well, such as replacing ' ' with '+'.  In addition, '=' is used as padding at the end of the encoding if the string isn't divisible by 3.  This is all part of the spec, the same removal is being done on the 3rd party server otherwise the code challenges won't match, nor will there conversion of the raw value.

File: pkg/services/auth/internal/gitlab.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	u.Scheme = gitlabScheme
+	u.Host = gitlabHost
+	u.Path = "/oauth/authorize"
+
+	params := u.Query()
+	params.Set(gitlabClientIdKey, gitlabClientId)
+	params.Set(gitlabRedirectUriKey, redirectUri)
+	params.Set("response_type", gitLabCodeKey)
+	params.Set("code_challenge", verifier.CodeChallenge())
+	params.Set("code_challenge_method", "S256")
```
[the remark this answers is not in the document]
Comment (JamWils) [reply pkg/services/auth/internal/gitlab.go:45]: Honestly, the ones that are const were used more than once.  I can go ahead and switch them back to strings.  I almost didn't do it in the first place because I thought the difference looked weird too.

File: pkg/services/auth/internal/gitlab.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+	gitlabRedirectUriKey = "redirect_uri"
+	gitLabCodeKey        = "code"
+	gitlabScheme         = "https"
+	gitlabHost           = "gitlab.com"
+	gitlabClientId       = "a28c1cb975f2cffd749bc7e32a5a1ebce8a3e7a2812ff27b54d2265ea0de3633"
+	gitlabClientSecret   = "6f173cc1018f7e6cfdd76b1912dfd03b15aaf8a4cb3a5aad4b30a219170408eb"
+
+	GitlabVerifierMin    = 43
+	GitlabVerifierMax    = 128
+	GitlabRedirectUriCLI = "http://127.0.0.1:9999/oauth/gitlab/callback"
```
[the remark this answers is not in the document]
Comment (JamWils) [reply pkg/services/auth/internal/gitlab.go:19]: Yes, otherwise the app will not work.

File: pkg/services/auth/gitlab.go [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+		fmt.Fprintf(w, "Shutting the server down...\n")
+		if err := server.Shutdown(shutdownCtx); err != nil {
+			panic(err)
+		}
+
+		fmt.Fprintf(w, "Server shutdown complete!\n")
+
+		if token.Err != nil {
+			fmt.Fprintf(w, "There was an issue going through the Gitlab authentication flow:\n\n")
+			if token.HttpStatusCode == 0 || token.HttpStatusCode == http.StatusOK {
```
[the remark this answers is not in the document]
Comment (JamWils) [reply pkg/services/auth/gitlab.go:158]: If the callback flow is never actually reached.  E.g. Gitlab is down and we can't even make a connection to the client.

File: pkg/services/auth/gitlab.go [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+			serverShutdown.Done()
+			token.Err = fmt.Errorf("gitlab auth client error: %w", clientErr)
+		}
+
+		serverShutdown.Wait()
+		shutdownCtx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+		defer cancel()
+		fmt.Fprintf(w, "Shutting the server down...\n")
+		if err := server.Shutdown(shutdownCtx); err != nil {
+			panic(err)
```
[the remark this answers is not in the document]
Comment (JamWils) [reply pkg/services/auth/gitlab.go:151]: Nope, that is a mistake, good catch.  I'll fix it.

File: pkg/services/auth/gitlab.go [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+		res, err := gaf.client.Do(req)
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			w.Write([]byte(http.StatusText(http.StatusInternalServerError)))
+			tokenState.HttpStatusCode = http.StatusInternalServerError
+			tokenState.Err = fmt.Errorf("gitlab token requeset client issue: %w", err)
+			return
+		}
+
+		switch res.StatusCode {
```
[the remark this answers is not in the document]
Comment (JamWils) [reply pkg/services/auth/gitlab.go:82]: My intent was to handle some other status codes specifically, but I couldn't get much info from their documentation so I just handled 200 and never reverted this to an if-statement.  I'll fix it.

File: pkg/services/auth/gitlab.go [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+		}
+
+		fmt.Fprintf(w, "Server shutdown complete!\n")
+
+		if token.Err != nil {
+			fmt.Fprintf(w, "There was an issue going through the Gitlab authentication flow:\n\n")
+			if token.HttpStatusCode == 0 || token.HttpStatusCode == http.StatusOK {
+				fmt.Fprintf(w, "%s", token.Err.Error())
+			} else {
+				fmt.Fprintf(w, "Gitlab returned status code %d with the error message: %s", token.HttpStatusCode, token.Err.Error())
```
[the remark this answers is not in the document]
Comment (JamWils) [reply pkg/services/auth/gitlab.go:161]: 🎉

File: pkg/services/auth/gitlab_test.go [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+		"https://weave.works/unit/test\n\n",
+		"Waiting for authentication flow completion...\n\n",
+		"Shutting the server down...\n",
+		"Server shutdown complete!\n",
+	}
+
+	return strings.Join(cliOutputSuccess, "")
+}
+
+var _ = FDescribe("Shutdown server handler", func() {
```
[the remark this answers is not in the document]
Comment (JamWils) [reply pkg/services/auth/gitlab_test.go:186]: 🤦 good catch.

File: pkg/services/auth/auth_suite_test.go [hunk 11]
```diff
@@ -1,11 +1,11 @@
 package auth
 
 import (
+	"github.com/weaveworks/weave-gitops/pkg/testutils"
 	"testing"
 
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
-	"github.com/weaveworks/weave-gitops/pkg/testutils"
```
[the remark this answers is not in the document]
Comment (JamWils) [reply pkg/services/auth/auth_suite_test.go:8]: Probably, there are different variations we can have.  What is the normal pattern that we use?  I can make sure my IDE is set up to follow that standard.

File: pkg/services/auth/gitlab.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (JamWils) [reply pkg/services/auth/gitlab.go]: Done

File: pkg/services/auth/gitlab.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (JamWils) [reply pkg/services/auth/gitlab.go]: Done! Can you double-check that you are good with the continued flow?

File: pkg/services/auth/gitlab_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (JamWils) [reply pkg/services/auth/gitlab_test.go]: Done, I did this in several places and they are all updated.

File: pkg/services/auth/internal/gitlab.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (JamWils) [reply pkg/services/auth/internal/gitlab.go:42]: Done

File: pkg/services/auth/auth_suite_test.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (jpellizzari) [reply pkg/services/auth/auth_suite_test.go:8]: I'm using `goimports` and it seems to not clobber everyone else

File: pkg/services/auth/gitlab.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (jpellizzari) [reply pkg/services/auth/gitlab.go:160]: You will have gotten a TCP error or a 502 from a Gitlab edge server somewhere earlier, but :shrug:

File: pkg/services/auth/internal/gitlab.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (jpellizzari) [reply pkg/services/auth/internal/gitlab.go:17]: Added

File: pkg/utils/string_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jpellizzari) [reply pkg/utils/string_test.go]: What mistake would someone have to make to see this panic show up in development?

File: pkg/gitproviders/factory.go [hunk 12]
```diff
@@ -30,13 +31,11 @@ type Config struct {
 	Token string
 }
 
-func buildGitProvider(config Config) (gitprovider.Client, error) {
+func buildGitProvider(config Config) (gitprovider.Client, string, error) {
```
Comment (jpellizzari) [inline pkg/gitproviders/factory.go:34]:
```suggestion
func buildGitProvider(config Config) (gitprovider.Client, url.URL, error) {
```
^^ Possible? A tad more expressive.

File: tools/download-deps.sh [hunk 13]
```diff
@@ -116,4 +116,7 @@ do_curl_binary "stoml" "${STOML_URL}" "${BIN_DIR}"
 tools=$("${BIN_DIR}"/stoml "${DEP_FILE}" .)
 for tool in $tools; do
     download_dependency "${tool}" "${BIN_DIR}"
-done
\ No newline at end of file
+done
+
+echo "Installing golangci-lint"
+curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(go env GOPATH)/bin v1.42.1
```
Comment (jpellizzari) [inline tools/download-deps.sh:122]: Does adding `golangci` to the `tools` package not auto-magically install the binary for you?

**Review by jpellizzari — APPROVED** [review]
LGTM. The Oauth callback URL has been added to the app.

File: pkg/utils/string_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (JamWils) [reply pkg/utils/string_test.go]: So in the case of the Gitlab example, I would pass the wrong numbers in the parameter.  When I go to test the Gitlab flow via unit tests or manually through the CLI it will panic.  It just can't happen.

File: tools/download-deps.sh (hunk 13 above)
Comment (JamWils) [reply tools/download-deps.sh:122]: The documentation recommended not to install it via `go get` but through a binary.  This seemed like the safer option, but I can add it as an import.

File: pkg/gitproviders/factory.go (hunk 12 above)
Comment (JamWils) [reply pkg/gitproviders/factory.go:34]: It's possible, but the receiving functions take a `string`.  I was trying to make the change as low friction as possible, it did make me sad to add an arbitrary return parameter on that function.

File: pkg/services/auth/auth_suite_test.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (JamWils) [reply pkg/services/auth/auth_suite_test.go]: Sigh, my 🧠 wasn't working this morning.  I could have just easily reverted this changed and realized I had moved this accidentally without realizing it.  The file is back to normal.

**Review by luizbafilho — APPROVED** [review]
